module/src/main/scala/com/gu/googleauth/auth.scala

```diff
@@ -63,8 +69,72 @@ object GoogleAuthConfig {
     maxAuthAge: Option[Duration] = defaultMaxAuthAge,
     enforceValidity: Boolean = defaultEnforceValidity,
     prompt: Option[String] = defaultPrompt,
-    antiForgeryKey: String = defaultAntiForgeryKey
-  ): GoogleAuthConfig = GoogleAuthConfig(clientId, clientSecret, redirectUrl, None, maxAuthAge, enforceValidity, prompt, antiForgeryKey)
+    antiForgeryChecker: AntiForgeryChecker
+  ): GoogleAuthConfig =
+    GoogleAuthConfig(clientId, clientSecret, redirectUrl, None, maxAuthAge, enforceValidity, prompt, antiForgeryChecker)
+}
+
+/**
+  * When the OAuth callback returns to our app, we need to ensure that this is the end of a valid authentication
+  * sequence that we initiated, and not a forged redirect. Rather than use a nonce, we use a signed session id
+  * in a short-lifetime Json Web Token, allowing us to cope better with concurrent authentication requests from the
+  * same browser session.
+  *
+  * "One good choice for a state token is a string of 30 or so characters constructed using a high-quality
+  * random-number generator. Another is a hash generated by signing some of your session state variables with
+  * a key that is kept secret on your back-end."
+  * - https://developers.google.com/identity/protocols/OpenIDConnect#createxsrftoken
+  *
+  * The design here is partially based on a IETF draft for "Encoding claims in the OAuth 2 state parameter ...":
+  * https://tools.ietf.org/html/draft-bradley-oauth-jwt-encoded-state-01
+  */
+case class AntiForgeryChecker(
+  stateSigningSecret: String,
+  jwtSignatureAlgorithm: SignatureAlgorithm,
+  sessionIdKeyName: String = "play-googleauth-session-id"
+) {
+
+  private val base64EncodedSecret: String =
+    Base64.getEncoder.encodeToString(stateSigningSecret.getBytes(UTF_8))
+
+  def ensureUserHasSessionId(t: String => Future[Result])(implicit request: RequestHeader, ec: ExecutionContext):Future[Result] = {
+    val sessionId = request.session.get(sessionIdKeyName).getOrElse(generateSessionId())
+
+    t(sessionId).map(_.addingToSession(sessionIdKeyName -> sessionId))
+  }
+
+  def generateToken(sessionId: String)(implicit clock: Clock = Clock.systemUTC) : String = {
+    Jwts.builder()
+      .setExpiration(Date.from(clock.instant().plusSeconds(60)))
+      .claim(SessionIdJWTClaimPropertyName, sessionId)
+      .signWith(jwtSignatureAlgorithm, base64EncodedSecret)
+      .compact()
```

Comment:
Here we're using the [JJWT](https://github.com/jwtk/jjwt) library to sign and later validate our JSON Web Tokens. This the same library used by Play to sign the `PLAY_SESSION` cookie since Play 2.6 (see https://github.com/playframework/playframework/pull/7053), so we don't need to add any additional dependencies to use it.